### PR TITLE
Use a `_recovery_source` if source is omitted or modified

### PR DESCRIPTION
**Pull request by s1monw**
Today if a user omits the `_source` entirely or modifies the source
on indexing we have no chance to re-create the document after it has
been added. This is an issue for CCR and recovery based on soft deletes
which we are going to make the default. This change adds an additional
recovery source if the source is disabled or modified that is only kept
around until the document leaves the retention policy window.

This change adds a merge policy that efficiently removes this extra source
on merge for all document that are live and not in the retention policy window
anymore.

it's not fully tested and needs some cleanups but I wanted to put it out here for discussion.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+        @Override
+        public DocValuesProducer getDocValuesReader() {
+            DocValuesProducer docValuesReader = super.getDocValuesReader();
+            return new FilterDocValuesProducer(docValuesReader) {
+                @Override
+                public NumericDocValues getNumeric(FieldInfo field) throws IOException {
+                    NumericDocValues numeric = super.getNumeric(field);
+                    if (recoverySourceField.equals(field.name)) {
+                        return new FilterNumericDocValues(numeric) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java]: maybe leave comments about why we don't need to check whether `numeric` and `docValuesReader` are `null`

**Review by jpountz — APPROVED** [review]
I like the approach.

Since I'd expect most documents to have dropped their recovery source already I'm wondering whether it would be a bit more efficient to compute the bitset of documents for which we need to retain the recovery source rather than drop.

File: server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java [hunk 2]
```diff
@@ -293,5 +298,15 @@ boolean isTombstone(int segmentDocId) throws IOException {
             }
             return tombstoneDV.advanceExact(segmentDocId) && tombstoneDV.longValue() > 0;
         }
+
+        boolean hasRecoverySource(int segmentDocId) throws IOException {
+            if (recoverySource == null) {
+                return false;
+            }
+            if (tombstoneDV.docID() > segmentDocId) {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java]: tombstoneDV -> recoverySource

File: server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java [hunk 3]
```diff
@@ -293,5 +298,15 @@ boolean isTombstone(int segmentDocId) throws IOException {
...
+                recoverySource = leafReader.getNumericDocValues(SourceFieldMapper.RECOVERY_SOURCE_NAME);
+            }
+            return tombstoneDV.advanceExact(segmentDocId);
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java]: Same here `tombstoneDV` -> `recoverySource`

File: server/src/main/java/org/elasticsearch/index/mapper/SourceFieldMapper.java [hunk 4]
```diff
@@ -240,8 +243,17 @@ protected void parseCreateField(ParseContext context, List<IndexableField> field
...
             BytesRef ref = source.toBytesRef();
             fields.add(new StoredField(fieldType().name(), ref.bytes, ref.offset, ref.length));
+        } else {
+            source = null;
         }
-    }
+
+        if (originalSource != null && source != originalSource && context.indexSettings().isSoftDeleteEnabled()) {
+            // if we omitted source or modified it we add the _recovery_source to ensure we have it for ops based recovery
+            BytesRef ref = source.toBytesRef();
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/mapper/SourceFieldMapper.java]: Should we store the original source here?

**Review by dnhatn — APPROVED** [review]
It's really neat :)

File: server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.BitSet;
+
+import java.io.IOException;
+import java.util.function.Supplier;
+
+final class RecoverySourcePruneMergePolicy extends OneMergeWrappingMergePolicy {
+    RecoverySourcePruneMergePolicy(String recoverySourceField, Supplier<Query> retentionPolicySupplier, MergePolicy in) {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java]: `retentionPolicySupplier` is confusing. It's a prune query supplier.

File: server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                    return wrapped;
+                }
+                return wrapReader(recoverySourceField, wrapped, retentionPolicySupplier);
+            }
+        });
+
+    }
+
+    // pkg private for testing
+    static CodecReader wrapReader(String recoverySourceField, CodecReader reader, Supplier<Query> retentionPolicySupplier)
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java]: Same here: `retentionPolicySupplier`

File: server/src/main/java/org/elasticsearch/index/mapper/SourceFieldMapper.java (hunk 4 above)
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/index/mapper/SourceFieldMapper.java]: LOL yeah we should :D - I will fix

**Comment by s1monw** [comment]
@dnhatn  @jpountz I addressed your comments and added more tests. Can you take another look?

File: server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        builder.add(new DocValuesFieldExistsQuery(recoverySourceField), BooleanClause.Occur.FILTER);
+        builder.add(retainSourceQuerySupplier.get(), BooleanClause.Occur.FILTER);
+        IndexSearcher s = new IndexSearcher(reader);
+        s.setQueryCache(null);
+        Weight weight = s.createWeight(builder.build(), false, 1.0f);
+        Scorer scorer = weight.scorer(reader.getContext());
+        if (scorer != null) {
+            return new SourcePruningFilterCodecReader(recoverySourceField, reader, BitSet.of(scorer.iterator(), reader.maxDoc()));
+        } else {
+            return new SourcePruningFilterCodecReader(recoverySourceField, reader, new BitSet.MatchNoBits(reader.maxDoc()));
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java]: s/BitSet.MatchNoBits/Bits.MatchNoBits/

File: server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+        @Override
+        public DocValuesProducer getDocValuesReader() {
+            DocValuesProducer docValuesReader = super.getDocValuesReader();
+            return new FilterDocValuesProducer(docValuesReader) {
+                @Override
+                public NumericDocValues getNumeric(FieldInfo field) throws IOException {
+                    NumericDocValues numeric = super.getNumeric(field);
+                    if (recoverySourceField.equals(field.name)) {
+                        assert numeric != null : recoverySourceField + " must have numeric DV but was null";
+                        return new FilterNumericDocValues(numeric) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java]: If `recoverySourceToKeep` was a bitset, we could do a leap frog, which would be faster if `recoverySourceToKeep` is sparse.

```java
final ConjunctionDISI intersection = ConjunctionDISI.intersectIterators(Arrays.asList(numeric, new BitSetIterator(recoverySourceToKeep)));
return new FilterNumericDocValues(numeric) {
  @Override
  public int nextDoc() throws IOException {
    return intersection.nextDoc();
  }
};
```

**Review by jpountz — APPROVED** [review]
LGTM. I left a suggestion for an improvement.

**Review by dnhatn — APPROVED** [review]
It's nice that we now use a single retention query for both MPs.

**Pull request merged**